### PR TITLE
[GreenCity] - Create new controller getUserAndCustomListByUserIdAndHabitId() #4763

### PR DESCRIPTION
**Pull request by NKorzhik**
Created new Create new controller getUserAndCustomListByUserIdAndHabitId() and test for it.

### CODE REVIEW

**Review by innayashna — APPROVED** [review]

File: core/src/test/java/greencity/controller/HabitAssignControllerTest.java [hunk 1]
```diff
@@ -210,4 +207,11 @@ void getUsersHabitByHabitId() throws Exception {
 
         verify(habitAssignService).findHabitByUserIdAndHabitId(null, 1L, "en");
     }
+
+    @Test
+    void getUserAndCustomListByUserIdAndHabitId() throws Exception {
+        mockMvc.perform(get(habitLink + "/allUserAndCustomList/{habitId}", 1L))
+            .andExpect(status().isOk());
+        verify(habitAssignService).getUserAndUserCustomShoppingList(null, 1L, "en");
```
Comment (RomaMocherniuk) [inline core/src/test/java/greencity/controller/HabitAssignControllerTest.java]: please add a test if you send `language` as a query param

File: service-api/src/main/java/greencity/dto/habit/HabitAssignUserAndUserCustomShoppingListDto.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+import java.util.List;
+
+@NoArgsConstructor
+@AllArgsConstructor
+@EqualsAndHashCode
+@ToString
+@Getter
+@Setter
+@Builder
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/dto/habit/HabitAssignUserAndUserCustomShoppingListDto.java]: please use `@Data` annotation

File: service/src/main/java/greencity/service/ShoppingListItemServiceImpl.java [hunk 3]
```diff
@@ -252,11 +252,7 @@ public List<UserShoppingListItemResponseDto> saveUserShoppingListItems(Long user
     private HabitAssign getHabitAssignIfExist(Long userId, Long habitId) {
         Optional<HabitAssign> habitAssignOptional =
             habitAssignRepo.findByHabitIdAndUserId(habitId, userId);
-        if (habitAssignOptional.isPresent()) {
-            return habitAssignOptional.get();
-        } else {
-            throw new NotFoundException(ErrorMessage.HABIT_ASSIGN_NOT_FOUND_BY_ID);
-        }
+        return habitAssignOptional.orElseGet(HabitAssign::new);
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/ShoppingListItemServiceImpl.java]: please add a test if `!habitAssignOptional.isPresent()` and update java docs

Comment (RomaMocherniuk) [reply service/src/main/java/greencity/service/ShoppingListItemServiceImpl.java]: Or maybe is it better to return `Optional<HabitAssign>`?

File: service/src/test/java/greencity/service/HabitAssignServiceImplTest.java [hunk 4]
```diff
@@ -4,29 +4,31 @@
...
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.mockito.InjectMocks;
 import org.mockito.Mock;
 import org.mockito.junit.jupiter.MockitoExtension;
 import org.modelmapper.ModelMapper;
 
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.*;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/HabitAssignServiceImplTest.java]: please do not use `*` for imports

File: service/src/test/java/greencity/service/HabitAssignServiceImplTest.java [hunk 5]
```diff
@@ -387,6 +395,28 @@ void getAllHabitAssignsByUserIdAndStatusNotCancelled() {
...
+            Collections.singletonList(customShoppingListItemResponseDto());
+        List<UserShoppingListItemResponseDto> userShoppingListItemResponseDtos =
+            Collections.singletonList(getCustomUserShoppingListItemDto());
+        HabitAssignUserAndUserCustomShoppingListDto expected = HabitAssignUserAndUserCustomShoppingListDto
+            .builder()
+            .customShoppingListItemDto(customShoppingListItemResponseDtos)
+            .userShoppingListItemsDto(userShoppingListItemResponseDtos)
+            .build();
+
+        when(shoppingListItemService.getUserShoppingList(anyLong(), eq(1L), eq("en")))
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/HabitAssignServiceImplTest.java]: It is better to create `useId`, `habitId` and `language` variables and use them when you mock some behavior

File: service/src/main/java/greencity/service/HabitAssignServiceImpl.java [hunk 6]
```diff
@@ -493,6 +495,23 @@ public List<HabitAssignDto> getAllHabitAssignsByUserIdAndStatusAcquired(Long use
...
+     */
+
+    @Override
+    public HabitAssignUserAndUserCustomShoppingListDto getUserAndUserCustomShoppingList(Long userId, Long habitId,
+        String language) {
+        HabitAssignUserAndUserCustomShoppingListDto habitDtos = new HabitAssignUserAndUserCustomShoppingListDto();
+        habitDtos.setUserShoppingListItemsDto(shoppingListItemService.getUserShoppingList(userId, habitId, language));
+        habitDtos.setCustomShoppingListItemDto(
+            customShoppingListItemService.findAllAvailableCustomShoppingListItems(userId, habitId));
+        return habitDtos;
```
Comment (RomaMocherniuk) [inline service/src/main/java/greencity/service/HabitAssignServiceImpl.java]: please use builder

File: core/src/main/java/greencity/controller/HabitAssignController.java [hunk 7]
```diff
@@ -147,6 +147,31 @@ public ResponseEntity<List<HabitAssignDto>> getCurrentUserHabitAssignsByIdAndAcq
...
+     * Method that return UserShoppingList and CustomShoppingList.
+     *
+     * @param habitId @link HabitVO} id.
+     * @param userVO  {@link UserVO} instance.
+     * @param locale  needed language code.
+     * @return User Shopping List and Custom Shopping List.
+     */
+    @ApiOperation(value = "Get user shopping and user custom shopping lists")
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = HttpStatuses.OK, response = HabitAssignDto.class,
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/HabitAssignController.java]: please update it
this endpoint does not return `List` of `HabitAssignDto`

File: core/src/main/java/greencity/controller/HabitAssignController.java [hunk 8]
```diff
@@ -147,6 +147,31 @@ public ResponseEntity<List<HabitAssignDto>> getCurrentUserHabitAssignsByIdAndAcq
                 .getAllHabitAssignsByUserIdAndStatusNotCancelled(userVO.getId(), locale.getLanguage()));
     }
 
+    /**
+     * Method that return UserShoppingList and CustomShoppingList.
+     *
+     * @param habitId @link HabitVO} id.
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/HabitAssignController.java]: `{@link HabitVO}`

File: service/src/test/java/greencity/service/HabitAssignServiceImplTest.java [hunk 9]
```diff
@@ -52,6 +54,12 @@ class HabitAssignServiceImplTest {
     @InjectMocks
     HabitAssignServiceImpl habitAssignService;
 
+    @Mock
+    ShoppingListItemService shoppingListItemService;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/HabitAssignServiceImplTest.java]: can be `private`

File: service/src/test/java/greencity/service/HabitAssignServiceImplTest.java [hunk 10]
```diff
@@ -52,6 +54,12 @@ class HabitAssignServiceImplTest {
...
+
+    @Mock
+    CustomShoppingListItemService customShoppingListItemService;
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/HabitAssignServiceImplTest.java]: same here

File: service/src/test/java/greencity/service/ShoppingListItemServiceImplTest.java [hunk 11]
```diff
@@ -414,16 +414,6 @@ void getUserShoppingListItemTest() {
         assertEquals(1L, shoppingListItemService.getUserShoppingList(userId, 1L, "en").get(0).getId());
     }
 
-    @Test
-    void getUserShoppingListItemsTestTrows() {
```
Comment (RomaMocherniuk) [inline service/src/test/java/greencity/service/ShoppingListItemServiceImplTest.java:418]: Please add a test if `userShoppingListItemRepo.findAllByHabitAssingId()` returns empty list

File: core/src/main/java/greencity/controller/HabitAssignController.java [hunk 12]
```diff
@@ -147,6 +147,31 @@ public ResponseEntity<List<HabitAssignDto>> getCurrentUserHabitAssignsByIdAndAcq
...
+     * @param userVO  {@link UserVO} instance.
+     * @param locale  needed language code.
+     * @return User Shopping List and Custom Shopping List.
+     */
+    @ApiOperation(value = "Get user shopping and user custom shopping lists")
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = HttpStatuses.OK, response = HabitAssignDto.class,
+            responseContainer = "List"),
+        @ApiResponse(code = 400, message = HttpStatuses.BAD_REQUEST),
+        @ApiResponse(code = 401, message = HttpStatuses.UNAUTHORIZED),
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/controller/HabitAssignController.java]: This endpoint is available for all user roles?
Or we could have `403` response too?

File: service/src/test/java/greencity/service/HabitAssignServiceImplTest.java (hunk 4 above)
Comment (NKorzhik) [reply service/src/test/java/greencity/service/HabitAssignServiceImplTest.java]: Is it bad practise to use `*` for imports or it is connected only java.util?

**Review by OlegVat — APPROVED** [review]

**Review by JJJazl — APPROVED** [review]

File: service-api/src/main/java/greencity/service/HabitAssignService.java [hunk 13]
```diff
@@ -105,6 +103,18 @@ HabitAssignUserShoppingListItemDto updateUserShoppingItemListAndDuration(Long ha
      */
     List<HabitAssignDto> getAllHabitAssignsByUserIdAndStatusNotCancelled(Long userId, String language);
 
+    /**
+     * Method that finds userShoppingListItems and userCustomShoppingListItems for
+     * habitId.
+     *
+     * @param habitId  {@code Habit} id.
+     * @param userId   {@code User} id.
```
Comment (RomaMocherniuk) [inline service-api/src/main/java/greencity/service/HabitAssignService.java:110]: please swap 110 and 111 lines
UserId is the first param

File: core/src/main/java/greencity/config/SecurityConfig.java [hunk 14]
```diff
@@ -27,7 +27,11 @@
 import org.springframework.web.cors.CorsConfigurationSource;
 import org.springframework.web.cors.UrlBasedCorsConfigurationSource;
 
-import static greencity.constant.AppConstant.*;
+import static greencity.constant.AppConstant.ADMIN;
+import static greencity.constant.AppConstant.USER;
+import static greencity.constant.AppConstant.MODERATOR;
+import static greencity.constant.AppConstant.UBS_EMPLOYEE;
+
```
Comment (RomaMocherniuk) [inline core/src/main/java/greencity/config/SecurityConfig.java:34]: 👍

**Review by RomaMocherniuk — APPROVED** [review]
LGTM

**Pull request merged**